### ISSUE TITLE
perf: DCGAN training perf plan — autograd alloc, Conv2D small-N, training pre-pack

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#403]
**Related:** ooples/AiDotNet#1390 (DCGAN training 5-10× slower than PyTorch CPU; the `DCGANTests.MoreData_ShouldNotDegrade` test times out at 250 iters)

## Why this is a Tensors issue

`AiDotNet.GenerativeAdversarialNetwork.Train()` orchestrates the training loop but ~95% of wall time per step lives in this repo:

- `Conv2DWithIm2ColDouble` (`Engines/CpuEngine.cs:8793`) — Im2Col + GEMM per batch slice
- `Im2ColHelper.Im2Col`
- `BlasProvider.TryGemm` → OpenBLAS or BlasManaged
- `DifferentiableOps.RecordBinary` + `GradNode` allocation (`Engines/Autodiff/`)
- All backward functions (`Engines/Autodiff/BackwardFunctions.cs`)
- `CompiledBackwardGraph` / `CompiledBackwardWalk` / `CompiledDelegateChain`
- Per-op tape recording closure capture

## Goal

≥3× speedup on the DCGAN<double> training step (519 ms → ≤170 ms) so `DCGANTests.MoreData_ShouldNotDegrade` clears 120 s budget at 250 iters.

## Substep profile from AiDotNet#1390

| Substep | Time | % |
|---|---:|---:|
| Generator.Predict (no-grad fwd) | 19.3 ms | 3.7% |
| Discriminator.Train (fwd + bwd + optim, batch=2) | 186.5 ms | 35.9% |
| Generator adversarial (tape gen+disc fwd + bwd + optim) | 313.3 ms | 60.4% |
| **Total** | **519 ms / step** | |

### Phase A — Investigation (1-2 days, prereq for B-E)

Without this no one knows where the time really goes. Three measurements:

1. **Shape-catalog dump.** Reuse `ShapeInstrumenter` + `BlasProvider.ShapeLogHook` (exists from Sub-A) to record every (M, N, K) GEMM driven by a 1-step DCGAN run. Output a catalog of unique shapes + call counts.
2. **Per-component microbench.** Stopwatch around: Im2Col, GEMM-forward, GEMM-backward-gradInput, GEMM-backward-gradWeight, `RecordBinary`, optimizer `Step()`, BatchNorm forward, BatchNorm backward.
3. **Allocation profile.** Run one step under `dotnet-trace collect --providers Microsoft-Windows-DotNETRuntime:0x1:5`. Identify top allocators. Hypothesis: `GradNode` + `BackwardFunction` closures + per-call im2col buffer.

**Files:** new `tests/AiDotNet.Tensors.Benchmarks/DCGANStepProbe.cs`; reuse `src/AiDotNet.Tensors/Engines/BlasManaged/Autotune/ShapeInstrumenter.cs`.

**Output:** short markdown report ranking the bottlenecks. Phases B-E are sequenced based on findings.

### Phase B — Autograd allocation cleanup (~10-30%, ~50-150 ms/step)

Addresses the issue's "per-op tape record + GradNode allocation" complaint.

1. **GradNode pooling per-tape.** `GradientTape<T>` already has a lifetime; add a `Stack<GradNode<T>>` field and recycle on `Dispose`. Each `RecordBinary` pulls from the pool instead of `new GradNode<T>(...)`. 11 layers × 2 dirs × 250 iters = 11,000 allocs per test.
2. **Closure-free backward dispatch.** Audit every `RecordBinary` call site for captured locals — `BackwardFunctions<T>` methods are static and should pass as method-group references (no allocation). Look for patterns like `var captured = ...; tape.Record(..., (g, ...) => UseCaptured(captured, ...))` and lift to no-capture form (save state in GradNode).
3. **Inline small-op backward.** `CompiledBackwardWalk.cs` already inlines some ops (per commit history). Extend to cover DCGAN's chain — BatchNorm, Conv2D backward, LeakyReLU.

**Files:**
- `src/AiDotNet.Tensors/Engines/Autodiff/GradNode.cs`
- `src/AiDotNet.Tensors/Engines/Autodiff/GradientTape.cs`
- `src/AiDotNet.Tensors/Engines/Autodiff/DifferentiableOps.cs`
- `src/AiDotNet.Tensors/Engines/Autodiff/BackwardFunctions.cs` (audit for captures)
- `src/AiDotNet.Tensors/Engines/Autodiff/CompiledBackwardWalk.cs` (extend inliners)

**Risk:** Medium — tape semantics interact with higher-order AD; tests must keep passing.

### Phase C — Conv2DBackward small-N kernel (~30-50%, ~150-260 ms/step)

Biggest single prize. `Conv2DWithIm2ColDouble`:
- Allocates `colH * colW = (inC × kH × kW) × (outH × outW)` per batch slice
- Calls `Im2Col` per batch slice (serial)
- Calls `BlasProvider.TryGemm` per batch slice (serial)

For DCGAN at batch=2, kernel=4×4, in=64, out=64, spatial=8×8:
- `colH = 1024`, `colW = 64`, GEMM shape M=64×N=64×K=1024
- Im2col buffer: 512 KB/slice
- Total work per layer: 16.7 MFLOPs forward — per-call dispatch + im2col overhead dominates at this scale

**Likely fixes:**
1. **Batched im2col + single GEMM.** Concatenate im2col for all batch slices once, then one big GEMM. One OpenBLAS dispatch instead of `batch`. At batch≥8 it dominates; at batch=2 still measurable. ~1 day to implement.
2. **Direct convolution loop for tiny spatial.** 1×1 bypass exists (line 8816). Add a similar bypass for `outH * outW < 256` — the late-DCGAN-decoder pattern. Direct loop avoids the im2col buffer entirely.
3. **Specialised deconv 4×4 stride=2 kernel.** Hard-codes the access pattern. `NchwcConv2D*.cs` already has FP32 NCHWc — extend to FP64 and/or deconv-stride-2.
4. **Reusable per-thread im2col buffer.** `[ThreadStatic]` buffer sized to largest seen shape — eliminates rent/return per call.

**Files:**
- `src/AiDotNet.Tensors/Engines/CpuEngine.cs` (Conv2DWithIm2ColDouble + Float twin)
- `src/AiDotNet.Tensors/Helpers/Im2ColHelper.cs`
- `src/AiDotNet.Tensors/Engines/Simd/NchwcConv2D.cs` / `NchwcConv2D16.cs`
- `src/AiDotNet.Tensors/Helpers/DirectConv2DHelper.cs` (check if DCGAN qualifies)

**Risk:** Medium-high. Conv2D edge cases (padding, stride, dilation) finicky; broad correctness tests required against a naive reference before perf claim.

### Phase D — TrainingWeightRegistry (~3-8%, ~15-40 ms/step)

Builds on Sub-E (PR #402). `FrozenWeightRegistry` is inference-only. Training analog:

```csharp
public static class TrainingWeightRegistry
{
    public static void Register<T>(Tensor<T> weight);     // pre-pack once
    public static WeightPackHandle? TryGetHandle<T>(Tensor<T>);
    public static void OnOptimizerStep<T>(Tensor<T>);     // MarkDirty + re-pack
}
```

Optimizer integration: in the param-update path, call `TrainingWeightRegistry.OnOptimizerStep(p)` after `p ← p - lr * grad`. The next forward gets a freshly-packed B. Within one step, forward and backward (gradA = dC · Bᵀ) reuse the pack.

**Additional Sub-E extension required:** also pre-pack Bᵀ for backward gradA computation. `FrozenWeightRegistry` currently stores only the forward-direction pre-pack.

**Files:**
- New `src/AiDotNet.Tensors/Engines/BlasManaged/TrainingWeightRegistry.cs`
- `src/AiDotNet.Tensors/Engines/Autodiff/BackwardFunctions.cs` — MatMulBackward fast path consults registry for Bᵀ pre-pack
- Optimizer side (verify location in Phase A)

**Risk:** Low — Sub-E infrastructure covers correctness; this is wiring.

### Phase E — FP64 microkernel quality

Sub-E touched FP32 scalar accuracy but not FP64 perf. Check what fires during DCGAN — small-K shapes may pick Streaming or PackAOnly and skip `Avx2Fp64_4x8`.

**Investigation:** hook `BlasManagedAutotune` to record strategy + microkernel per shape in Phase A.

**Possible fixes:**
- Verify FP64 tail-handling microkernels (`Avx2Tail.cs`) actually fire
- Wire `Avx512Fp64_8x16` if host supports (already wired — verify it fires)

**Risk:** Low — incremental tuning.

### Phase F — FP32 model variant (out of Tensors scope)

`DCGANTests.MoreData_ShouldNotDegrade_Fp32` parallel test. ~30-40% wall saving on AVX2 (8 lanes vs 4 for FP64). AiDotNet repo work — listed for completeness as the pragmatic escape hatch if A-E don't close the gap.

## Sequencing

1. **A first** (1-2 days, must precede everything)
2. **B** (1 week, allocation work — parallel to C)
3. **C** (1-2 weeks, kernel work — highest expected ROI per Phase A data)
4. **D** (2-3 days, after B+C stable so we can measure the additive win)
5. **E** (1 day, surgical tuning based on Phase A microkernel logs)

## Acceptance

`DCGANTests.MoreData_ShouldNotDegrade` clears 120 s on CI host with ≥20 s margin. Add a sentinel `[Fact]` that fails if mean step time exceeds 400 ms — catches future regression even before the timeout would.

## NOT this issue

- Frontend duplicate-forward (`Generator.Predict` runs twice per step in `GenerativeAdversarialNetwork.Train`). Tracked in AiDotNet#1390 Phase 1.
- DCGAN shape-contract bugs — fixed by AiDotNet#1389.
